File: app/src/androidTest/java/com/bbtest/perform/base/MemInfo.java
package com.bbtest.perform.base;


import androidx.test.uiautomator.UiDevice;

import com.bbtest.utils.ShellCommand;

public class MemInfo {

    private UiDevice device = null;
    private String pkgName = "";
    private long totalMem = 0;
    private long availabelMem = 0;
    private long totalPss = 0;
    private long dalvilkPss = 0;   //java heap
    private long nativePss = 0;      //native heap
    private long views = 0;
    private long webviews = 0;
    private long activities = 0;

    public MemInfo(UiDevice device, String pkgName) {
        this.device = device;
        this.pkgName = pkgName;
        readMeminfo();
        readRaminfo();
    }

    public long getTotalMem() {
        return totalMem;
    }

    public long getAvailabelMem() {
        return availabelMem;
    }

    public long getTotalPss() {
        return totalPss;
    }

    public long getDalvilkPss() {
        return dalvilkPss;
    }

    public long getNativePss() {
        return nativePss;
    }

    public long getViews() {
        return views;
    }

    public long getWebviews() {
        return webviews;
    }

    public long getActivities() {
        return activities;
    }

    private void readMeminfo() {
        String cmd = "dumpsys meminfo " + pkgName;
        String memInfo = ShellCommand.execCmdByUiDevice(device, cmd);
        String[] memInfoLines = memInfo.split("\n");
        for (String memInfoLine : memInfoLines) {
            memInfoLine = memInfoLine.trim();
            if (memInfoLine.startsWith("TOTAL") && !memInfoLine.contains("Size") && !memInfoLine
                    .contains("Free") && !memInfoLine.contains("PSS")) {
                totalPss = Long.parseLong(memInfoLine.split("\\s+")[1].trim());
            } else if (memInfoLine.startsWith("Dalvik") && memInfoLine.contains("Heap") && !memInfoLine
                    .contains(":")) {
                dalvilkPss = Long.parseLong(memInfoLine.split("\\s+")[2].trim());
            } else if (memInfoLine.startsWith("Native") && memInfoLine.contains("Heap") && !memInfoLine
                    .contains(":")) {
                nativePss = Long.parseLong(memInfoLine.split("\\s+")[2].trim());
            } else if (memInfoLine.startsWith("Views:")) {
                views = Long.parseLong(memInfoLine.split("\\s+")[1].trim());
            } else if (memInfoLine.startsWith("WebViews:")) {
                webviews = Long.parseLong(memInfoLine.split("\\s+")[1].trim());
            } else if (memInfoLine.startsWith("AppContexts:") && memInfoLine.contains("Activities:")) {
                activities = Long.parseLong(memInfoLine.split("\\s+")[3].trim());
            }
        }
    }

    private void readRaminfo() {
        String cmd = "cat /proc/meminfo";
        String ramInfo = ShellCommand.execCmdByUiDevice(device, cmd);
        String[] ramInfoLines = ramInfo.split("\n");
        for (String ramInfoLine : ramInfoLines) {
            ramInfoLine = ramInfoLine.trim();
            if (ramInfoLine.startsWith("MemTotal:")) {
                totalMem = Long.parseLong(ramInfoLine.trim().split("\\s+")[1].trim());
            } else if (ramInfoLine.startsWith("MemAvailable:") || ramInfoLine.startsWith("MemFree:")) {
                availabelMem = Long.parseLong(ramInfoLine.trim().split("\\s+")[1].trim());
            }
        }
    }

}

File: app/src/androidTest/kotlin/com/bbtest/perform/base/MemInfo.kt
package com.bbtest.perform.base

import androidx.test.uiautomator.UiDevice
import com.bbtest.utils.ShellCommand

class MemInfo(
    private val device: UiDevice,
    private val pkgName: String,
) {
    private var totalMem = 0L
    private var availabelMem = 0L
    private var totalPss = 0L
    private var dalvilkPss = 0L
    private var nativePss = 0L
    private var views = 0L
    private var webviews = 0L
    private var activities = 0L

    init {
        readMeminfo()
        readRaminfo()
    }

    fun getTotalMem(): Long = totalMem

    fun getAvailabelMem(): Long = availabelMem

    fun getTotalPss(): Long = totalPss

    fun getDalvilkPss(): Long = dalvilkPss

    fun getNativePss(): Long = nativePss

    fun getViews(): Long = views

    fun getWebviews(): Long = webviews

    fun getActivities(): Long = activities

    private fun readMeminfo() {
        val memInfo = ShellCommand.execCmdByUiDevice(device, "dumpsys meminfo $pkgName")
        val memInfoLines = memInfo.split("\n")
        for (line in memInfoLines) {
            val memInfoLine = line.trim()
            if (memInfoLine.startsWith("TOTAL") &&
                !memInfoLine.contains("Size") &&
                !memInfoLine.contains("Free") &&
                !memInfoLine.contains("PSS")
            ) {
                totalPss = memInfoLine.split("\\s+".toRegex())[1].trim().toLong()
            } else if (memInfoLine.startsWith("Dalvik") && memInfoLine.contains("Heap") && !memInfoLine.contains(":")) {
                dalvilkPss = memInfoLine.split("\\s+".toRegex())[2].trim().toLong()
            } else if (memInfoLine.startsWith("Native") && memInfoLine.contains("Heap") && !memInfoLine.contains(":")) {
                nativePss = memInfoLine.split("\\s+".toRegex())[2].trim().toLong()
            } else if (memInfoLine.startsWith("Views:")) {
                views = memInfoLine.split("\\s+".toRegex())[1].trim().toLong()
            } else if (memInfoLine.startsWith("WebViews:")) {
                webviews = memInfoLine.split("\\s+".toRegex())[1].trim().toLong()
            } else if (memInfoLine.startsWith("AppContexts:") && memInfoLine.contains("Activities:")) {
                activities = memInfoLine.split("\\s+".toRegex())[3].trim().toLong()
            }
        }
    }

    private fun readRaminfo() {
        val ramInfo = ShellCommand.execCmdByUiDevice(device, "cat /proc/meminfo")
        val ramInfoLines = ramInfo.split("\n")
        for (line in ramInfoLines) {
            val ramInfoLine = line.trim()
            if (ramInfoLine.startsWith("MemTotal:")) {
                totalMem = ramInfoLine.split("\\s+".toRegex())[1].trim().toLong()
            } else if (ramInfoLine.startsWith("MemAvailable:") || ramInfoLine.startsWith("MemFree:")) {
                availabelMem = ramInfoLine.split("\\s+".toRegex())[1].trim().toLong()
            }
        }
    }
}
